### PR TITLE
use poetry instead of setup tools as package manager

### PR DESCRIPTION
**Pull request by FabienArcellier**
### Check
implement #230 

Peek 2024-02-07 08-12

### Check

* [x] a user can run `streamsync hello` after installing the package with `pip install streamsync[ds]`
* [x] the user can still install optional dependencies `pip install streamsync[ds]`
* [x] a developper can run `bash ./build.sh`
* [x] files referenced in `MANIFEST.in` are present in the distribution package

### Todo

* [x] configure poetry
* [x] configure poetry to handle virtualenv in `.venv`
* [x] migrate the build.sh
* [x] remove MANIFEST.in because poetry handle all the files from src
* [x] modify CONTRIBUTING.md
* [x] remove the dependency group of `test`, use `build` instead

### CODE REVIEW

File: src/streamsync/app_runner.py [hunk 1]
```diff
@@ -539,7 +539,7 @@ def __init__(self, app_path: str, mode: str):
         self.is_app_process_server_ready = multiprocessing.Event()
         self.is_app_process_server_failed = multiprocessing.Event()
         self.app_process_listener: Optional[AppProcessListener] = None
-        self.observer: Optional[watchdog.observers.Observer] = None
+        self.observer: Optional[PollingObserver] = None
```
Comment (ramedina86) [inline src/streamsync/app_runner.py:542]: Was this intentional?

File: src/streamsync/__init__.py [hunk 2]
```diff
@@ -1,8 +1,9 @@
+import importlib.metadata
 from typing import Union, Optional, Dict, Any
 from streamsync.core import Readable, FileWrapper, BytesWrapper, Config
 from streamsync.core import initial_state, component_manager, session_manager, session_verifier
 
-VERSION = "0.3.0"
+VERSION = importlib.metadata.version("streamsync")
```
Comment (ramedina86) [inline src/streamsync/__init__.py:6]: So the idea is to switch to defining version in the package and then fetching it from the code? Makes more sense I guess

File: src/streamsync/app_runner.py (hunk 1 above)
Comment (FabienArcellier) [reply src/streamsync/app_runner.py:542]: It is, otherwise `mypy` was complaining in `build.sh`

**Comment by ramedina86** [comment]
Looks good and thanks very much for updating the `CONTRIBUTING.md`, I'll have to consult that too.

@mmikita95 @raaymax Please have a look at this great example of updating docs while making changes in the code... Avoids documentation drift. I expect everyone to help keep docs updated; it's straightforward as it's just modifying the markdown files.

File: pyproject.toml [hunk 3]
```diff
@@ -1,59 +1,53 @@
...
+altair = ">= 5.2.0, < 6"
+httpx = ">=0.26.0, < 1"
 
-[project.scripts]
-streamsync = "streamsync.command_line:main"
+[tool.poetry.extras]
+ds = ["pandas", "pyarrow", "plotly"]
 
-[tool.setuptools.dynamic]
-version = { attr = "streamsync.VERSION" }
```
[the remark this answers is not in the document]
Comment (FabienArcellier) [reply pyproject.toml:59]: In the design choices of poetry, there is a strong choice not to read a python module dynamically like in setuptools. I understand this choice but it leads to a frustrating limit. I prefer the setuptools strategy but impossible to use it properly with poetry.

A [plugin written by the creator of fastapi](https://github.com/tiangolo/poetry-version-plugin) go over this limit but this plugin is incompatible with src layout and the PR to fix it has been open for 3 months and hasn't been merged.

File: src/streamsync/app_runner.py (hunk 1 above)
Comment (ramedina86) [reply src/streamsync/app_runner.py:542]: Ah nice, must be related to the dependencies I upgraded, thanks for fixing.

File: src/streamsync/__init__.py (hunk 2 above)
Comment (FabienArcellier) [reply src/streamsync/__init__.py:6]: That's the recommended way to manage this situation. Another one is to have a utility as bumpversion to update both location. I would avoid this one.

Comment (ramedina86) [reply src/streamsync/__init__.py:6]: I really like this option.

**Comment by mmikita95** [comment]
Indeed, beautifully done. I'll make sure to commit to that as well.

**Review by raaymax — COMMENTED** [review]
Nice one :)

**Pull request merged**